### ISSUE TITLE
Support all LightGBM parallel tree learners in Dask

### ISSUE THREAD

**Issue by StrikerRUS** [microsoft/LightGBM#3834]
According to this piece of code 
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/python-package/lightgbm/dask.py#L251-L264
Dask module supports [all available tree learner types](https://lightgbm.readthedocs.io/en/latest/Parameters.html#tree_learner).
Refer to https://lightgbm.readthedocs.io/en/latest/Parallel-Learning-Guide.html#choose-appropriate-parallel-algorithm for more details about difference among `tree_learner` param values.

However, our CI tests are run only for `data` parallel tree learner type.
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L182

I believe that parametrization of all tests with different tree learners will improve tests and make us confident in Dask module quality.

How tests can be parametrized:
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L31
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/tests/python_package_test/test_dask.py#L166

**Comment by jameslamb** [comment]
@StrikerRUS I disagree with calling this a "good first issue", and believe it should be a feature request. Using a different tree learner isn't as simple as just changing the tree_learner parameter. You have to be sure that the way that your data (either Dask DataFrame or Dask Array) is partitioned matches the method you chose.

For a dataset with `n` observations and `k`  columns:

* feature parallel = partitions have shape `(n, k_i)`. All rows but a subset of features.
* data parallel = partitions have shape `(n_i, k)`. All features but a subset of rows

Implementing this will require some non-trivial changes  to the tests and the Dask module.

**Comment by jameslamb** [comment]
I'm going to close this and add it to #2302. Thank you very much for writing this up, and for all of the Dask issues you've written up! Having this all documented in the backlog takes a lot of work but it's also critical to getting other contributors to come help.

**Comment by StrikerRUS** [comment]
> Implementing this will require some non-trivial changes to the tests and the Dask module.

Huh, I was suspecting this but shy to say it out loud! 🙂 

Given that only `data` tree learner is actually supported right now, I strongly believe we should make it explicit and raise errors for these values
https://github.com/microsoft/LightGBM/blob/ac706e10e461d1c5987a3975cc48aa82fde33566/python-package/lightgbm/dask.py#L254-L257

**Comment by jameslamb** [comment]
> I strongly believe we should make it explicit and raise errors for these values

I've proposed a PR: https://github.com/microsoft/LightGBM/pull/3848

**Comment by StrikerRUS** [comment]
Seems that `feature_parallel` cannot be supported
https://github.com/microsoft/LightGBM/blob/7880b79fde36873b95839c0029ea9d649397da51/src/c_api.cpp#L130-L131

**Comment by jameslamb** [comment]
Oh wow!

For context, looks like that comment was added in https://github.com/microsoft/LightGBM/pull/986.

So does this mean feature-parallel learning is only supported in the CLI?

**Comment by StrikerRUS** [comment]
> So does this mean feature-parallel learning is only supported in the CLI?

I think so.

**Comment by jameslamb** [comment]
ok I can make a separate pull request to update the parameter documentation for that, then

**Comment by jameslamb** [comment]
Adding some information for anyone looking to add `voting_parallel` support.

You can see a description at https://lightgbm.readthedocs.io/en/latest/Features.html#optimization-in-network-communication. And much more details in https://proceedings.neurips.cc/paper/2016/file/10a5ab2db37feedfdeaab192ead4ac0e-Paper.pdf if you're curious.

From the paper:

> Data-parallel: Training data are horizontally partitioned according to the samples and allocated to
different machines. Then the machines communicate with each other the local histograms of all
attributes (according to their own data samples) in order to obtain the global attribute distributions and
identify the best attribute and split point [12] [14]

> In this paper, we proposed a new data-parallel algorithm for decision tree, called Parallel Voting
Decision Tree (PV-Tree), which can achieve much better balance between communication efficiency
and accuracy. The key difference between conventional data-parallel decision tree algorithm and
PV-Tree lies in that the former only trusts the globally aggregated histogram information, while the
latter leverages the local statistical information contained in each machine through a two-stage voting
process, thus can significantly reduce the communication cost. Specifically, PV-Tree contains the
following steps in each iteration. 1) Local voting. On each machine, we select the top-k attributes
based on its local data according to the informativeness scores (e.g., risk reduction for regression,
and information gain for classification). 2) Global voting. We determine global top-2k attributes
by a majority voting among the local candidates selected in the previous step. That is, we rank the
attributes according to the number of local machines who select them, and choose the top 2k attributes
from the ranked list. 3) Best attribute identification. We collect the full-grained histograms of the
globally top-2k attributes from local machines in order to compute their global distributions.

In theory, adding `voting_parallel` to `lightgbm.dask` might be as simple as adding it to the unit tests (basically `pytest.mark.parametrize("tree_learner", distributed_training_algorithms)` on `test_classifier`, `test_regressor`, and `test_ranker` in https://github.com/microsoft/LightGBM/blob/4580393f604d825c318c053891f2870e6a40347f/tests/python_package_test/test_dask.py) and the altering this check:

https://github.com/microsoft/LightGBM/blob/4580393f604d825c318c053891f2870e6a40347f/python-package/lightgbm/dask.py#L274

Voting parallel uses horizontally-partitioned data just like `data_parallel`,  so none of the data distribution stuff in `lightgbm.dask` should need to change.

**Comment by jmoralez** [comment]
>     So does this mean feature-parallel learning is only supported in the CLI?
> 
> I think so.

Should I remove this? https://github.com/microsoft/LightGBM/blob/971b5486873339bcaa2bbe0948c95e1fff7246c7/python-package/lightgbm/dask.py#L265-L266

**Comment by jameslamb** [comment]
No, I'd prefer to leave that and rely on LightGBM to throw the error mentioned above.

If you try to do feature parallel and your data are vertically partitioned, it wouldn't mske sense for LightGBM to try to perform data parallel learning (which requires *horizontally partitioned* data).

**Comment by jmoralez** [comment]
Adding `voting` and `voting_parallel` here: https://github.com/microsoft/LightGBM/blob/971b5486873339bcaa2bbe0948c95e1fff7246c7/python-package/lightgbm/dask.py#L274 
Just leaves the feature parallel to go into the if and trigger this warning first: `UserWarning: Support for tree_learner feature_parallel in lightgbm.dask is experimental and may break in a future release.` and then the error. Maybe we should remove the if and go straight to the error? The warning kind of suggests that there is support for feature_parallel in the dask api but there isn't.

**Comment by jameslamb** [comment]
oh good point! Ok I would support removing that warning entirely at the same time you add support for voting parallel.
